In [51]:
# Read data sets on ETF and mutual funds 
# Identify a historical financial dataset including a measure of financial risk
import pandas as pd
etf = pd.read_csv("/Users/ishaanbabbar/6400-Project-Group2/data/ETFs.csv")
etf.head()

fund_symbol quote_type region                  fund_short_name  \
0        AAAU        ETF     US  DWS RREEF Real Assets Fund - Cl   
1        AADR        ETF     US  AllianzGI Health Sciences Fund    
2        AAXJ        ETF     US                              NaN   
3        ABEQ        ETF     US  Thrivent Large Cap Growth Fund    
4        ACES        ETF     US                              NaN   

                                      fund_long_name currency  \
0               DWS RREEF Real Assets Fund - Class A      USD   
1      Virtus AllianzGI Health Sciences Fund Class P      USD   
2  American Century One Choice Blend+ 2015 Portfo...      USD   
3             Thrivent Large Cap Growth Fund Class A      USD   
4  American Century One Choice Blend+ 2015 Portfo...      USD   

               fund_category                   fund_family exchange_code  \
0                        NaN                           DWS           PCX   
1       Foreign Large Growth                        Virtus           NGM   
2  Pacific/Asia ex-Japan Stk  American Century Investments           NGM   
3                Large Value                Thrivent Funds           PCX   
4       Miscellaneous Sector  American Century Investments           PCX   

  exchange_name  ... fund_stdev_5years  fund_sharpe_ratio_5years  \
0      NYSEArca  ...               NaN                       NaN   
1      NasdaqGM  ...             19.30                      0.62   
2      NasdaqGM  ...             15.91                      0.66   
3      NYSEArca  ...               NaN                       NaN   
4      NYSEArca  ...               NaN                       NaN   

   fund_treynor_ratio_5years  fund_alpha_10years  fund_beta_10years  \
0                        NaN                 NaN                NaN   
1                       9.66                3.32               0.96   
2                      10.37                0.30               0.99   
3                        NaN                 NaN                NaN   
4                        NaN                 NaN                NaN   

   fund_mean_annual_return_10years  fund_r_squared_10years  \
0                              NaN                     NaN   
1                             0.79                   73.64   
2                             0.55                   78.24   
3                              NaN                     NaN   
4                              NaN                     NaN   

   fund_stdev_10years  fund_sharpe_ratio_10years  fund_treynor_ratio_10years  
0                 NaN                        NaN                         NaN  
1               16.78                       0.53                        8.15  
2               16.83                       0.36                        4.81  
3                 NaN                        NaN                         NaN  
4                 NaN                        NaN                         NaN  

[5 rows x 142 columns]

In [52]:
#Train and test several network models with the historical data
# Perform Data pre-processing and remove features in the dataset that aren't needed 
# Many columns have a higher percentage or missing columns, we will remove unnecessary columns 
# Check the percentage of missing values in each column 
missing_perc = etf.isnull().mean() * 100

# Show the columns with a higher than 50 percent missing values and drop them
missing_columns = missing_perc[missing_perc > 50]
etf_cleaned = etf.drop(columns = missing_columns.index)

# Fill the remaining missing values with median to retain integrity in the data and make it numeric
numeric_cols = etf_cleaned.select_dtypes(include = ['float64', 'int64']).columns
etf_cleaned[numeric_cols] = etf_cleaned[numeric_cols].fillna(etf_cleaned[numeric_cols].median())
etf_cleaned.head()

fund_symbol quote_type region                  fund_short_name  \
0        AAAU        ETF     US  DWS RREEF Real Assets Fund - Cl   
1        AADR        ETF     US  AllianzGI Health Sciences Fund    
2        AAXJ        ETF     US                              NaN   
3        ABEQ        ETF     US  Thrivent Large Cap Growth Fund    
4        ACES        ETF     US                              NaN   

                                      fund_long_name currency  \
0               DWS RREEF Real Assets Fund - Class A      USD   
1      Virtus AllianzGI Health Sciences Fund Class P      USD   
2  American Century One Choice Blend+ 2015 Portfo...      USD   
3             Thrivent Large Cap Growth Fund Class A      USD   
4  American Century One Choice Blend+ 2015 Portfo...      USD   

               fund_category                   fund_family exchange_code  \
0                        NaN                           DWS           PCX   
1       Foreign Large Growth                        Virtus           NGM   
2  Pacific/Asia ex-Japan Stk  American Century Investments           NGM   
3                Large Value                Thrivent Funds           PCX   
4       Miscellaneous Sector  American Century Investments           PCX   

  exchange_name  ... category_return_2009  category_return_2008  \
0      NYSEArca  ...              0.32673              -0.39004   
1      NasdaqGM  ...              0.29080              -0.42427   
2      NasdaqGM  ...              0.73761              -0.51342   
3      NYSEArca  ...              0.25157              -0.35846   
4      NYSEArca  ...              0.31262              -0.33986   

   category_return_2007  fund_alpha_3years  fund_beta_3years  \
0               0.08189             13.180              0.07   
1               0.31193             -1.300              1.11   
2               0.39069              1.200              0.90   
3               0.01809             -0.715              1.02   
4               0.01815             24.180              1.31   

   fund_mean_annual_return_3years  fund_r_squared_3years  fund_stdev_3years  \
0                            1.23                   0.54              14.93   
1                            0.85                  75.96              22.42   
2                            0.80                  74.34              18.48   
3                            0.88                  78.01              19.66   
4                            3.65                  52.02              32.60   

   fund_sharpe_ratio_3years  fund_treynor_ratio_3years  
0                      0.91                     187.10  
1                      0.40                       6.11  
2                      0.46                       7.80  
3                      0.56                       8.40  
4                      1.30                      34.50  

[5 rows x 85 columns]

In [3]:
# Convert etf clened data into a csv
etf_cleaned = etf_cleaned.drop(columns = ['investment_strategy'])
etf_cleaned.to_csv("/Users/ishaanbabbar/Desktop/GT Summer 2024/DSAN 6400/project/ETF_cleaned.csv")

In [71]:
# Identify columns that are focused on fund returns for the correlation matrix and construct the network with the established edges between nodes

additional_cols = [
    'fund_annual_report_net_expense_ratio', 'fund_price_book_ratio', 'fund_price_cashflow_ratio',
    'fund_price_earning_ratio', 'fund_price_sales_ratio', 'fund_sharpe_ratio_3years',
    'fund_treynor_ratio_3years', 'fund_beta_3years', 'fund_r_squared_3years',
    'fund_stdev_3years', 'fund_sharpe_ratio_3years', 'fund_treynor_ratio_3years', 'fund_return_ytd', 
    'fund_return_1month', 'fund_return_3months', 'fund_return_1year', 'fund_return_3years', 'fund_return_2020',
    'fund_return_2019', 'fund_return_2018', 'fund_return_2017'
]

cen_data = etf_cleaned[additional_cols]


# Ensure the columns are numeric
return_columns = [col for col in cen_data.columns if 'fund' in col and 'category' not in col]

#Fully ensure all areas are numeric 
return_columns = [col for col in return_columns if pd.api.types.is_numeric_dtype(etf_cleaned[col])]

# Extract the return values 
return_data = etf_cleaned[return_columns].copy()

# Calculate correlation matrix for fund returns 
return_matrix = return_data.corr()

return_matrix

fund_annual_report_net_expense_ratio  \
fund_annual_report_net_expense_ratio                              1.000000   
fund_price_book_ratio                                             0.017880   
fund_price_cashflow_ratio                                         0.046650   
fund_price_earning_ratio                                         -0.044600   
fund_price_sales_ratio                                            0.016170   
fund_sharpe_ratio_3years                                         -0.275047   
fund_treynor_ratio_3years                                         0.005918   
fund_beta_3years                                                 -0.052566   
fund_r_squared_3years                                            -0.072356   
fund_stdev_3years                                                 0.308247   
fund_sharpe_ratio_3years                                         -0.275047   
fund_treynor_ratio_3years                                         0.005918   
fund_return_ytd                                                   0.061808   
fund_return_1month                                                0.034467   
fund_return_3months                                               0.030360   
fund_return_1year                                                 0.060892   
fund_return_3years                                               -0.194871   
fund_return_2020                                                 -0.129757   
fund_return_2019                                                 -0.073059   
fund_return_2018                                                 -0.092000   
fund_return_2017                                                 -0.028512   

                                      fund_price_book_ratio  \
fund_annual_report_net_expense_ratio               0.017880   
fund_price_book_ratio                              1.000000   
fund_price_cashflow_ratio                          0.801040   
fund_price_earning_ratio                           0.709158   
fund_price_sales_ratio                             0.555176   
fund_sharpe_ratio_3years                           0.228188   
fund_treynor_ratio_3years                          0.008641   
fund_beta_3years                                  -0.000941   
fund_r_squared_3years                              0.082559   
fund_stdev_3years                                 -0.035833   
fund_sharpe_ratio_3years                           0.228188   
fund_treynor_ratio_3years                          0.008641   
fund_return_ytd                                   -0.046723   
fund_return_1month                                 0.280450   
fund_return_3months                                0.104084   
fund_return_1year                                  0.044516   
fund_return_3years                                 0.207095   
fund_return_2020                                   0.316999   
fund_return_2019                                   0.140507   
fund_return_2018                                   0.134646   
fund_return_2017                                   0.082522   

                                      fund_price_cashflow_ratio  \
fund_annual_report_net_expense_ratio                   0.046650   
fund_price_book_ratio                                  0.801040   
fund_price_cashflow_ratio                              1.000000   
fund_price_earning_ratio                               0.743135   
fund_price_sales_ratio                                 0.670423   
fund_sharpe_ratio_3years                               0.210015   
fund_treynor_ratio_3years                              0.010518   
fund_beta_3years                                      -0.021600   
fund_r_squared_3years                                  0.053885   
fund_stdev_3years                                     -0.055364   
fund_sharpe_ratio_3years                               0.210015   
fund_treynor_ratio_3years                              0.010518   
fund_return_ytd                                       -0.088159

In [77]:
# Calcualte the degree, betweenness and closeness centralities 
#Evaluate measures of centrality (betweenness, degree, closeness)
import numpy as np
import networkx as nx 
# Create a new network graph
G = nx.Graph()

# Add nodes(return funds in this context)
for funds in return_columns:
    G.add_node(funds)

# Add edges based on given threshold 
threshold = 0.2 
for i in range(len(return_matrix)):
    for j in range(i+1, len(return_matrix)):
        if abs(return_matrix.iloc[i,j]) > threshold:
            G.add_edge(return_matrix.index[i], return_matrix.index[j], weight = return_matrix.iloc[i,j])

# Compute centrality measures 
degree_centrality = nx.degree_centrality(G)
betweeness_centrality = nx.betweenness_centrality(G)
closeness_centrality = nx.closeness_centrality(G)


#Compare model performance and determine best model
centrality_df = pd.DataFrame({
    "Degree Centrality": degree_centrality,
    "Betweenness Centrality": betweeness_centrality,
    "Closeness Centrality": closeness_centrality
})
centrality_df

Degree Centrality  \
fund_annual_report_net_expense_ratio           0.111111   
fund_price_book_ratio                          0.388889   
fund_price_cashflow_ratio                      0.333333   
fund_price_earning_ratio                       0.333333   
fund_price_sales_ratio                         0.166667   
fund_sharpe_ratio_3years                       0.777778   
fund_treynor_ratio_3years                      0.111111   
fund_beta_3years                               0.500000   
fund_r_squared_3years                          0.111111   
fund_stdev_3years                              0.333333   
fund_return_ytd                                0.444444   
fund_return_1month                             0.333333   
fund_return_3months                            0.444444   
fund_return_1year                              0.611111   
fund_return_3years                             0.611111   
fund_return_2020                               0.500000   
fund_return_2019                               0.555556   
fund_return_2018                               0.388889   
fund_return_2017                               0.388889   

                                      Betweenness Centrality  \
fund_annual_report_net_expense_ratio                0.000000   
fund_price_book_ratio                               0.058397   
fund_price_cashflow_ratio                           0.027135   
fund_price_earning_ratio                            0.027135   
fund_price_sales_ratio                              0.000000   
fund_sharpe_ratio_3years                            0.179910   
fund_treynor_ratio_3years                           0.000000   
fund_beta_3years                                    0.014070   
fund_r_squared_3years                               0.000000   
fund_stdev_3years                                   0.028994   
fund_return_ytd                                     0.021322   
fund_return_1month                                  0.045985   
fund_return_3months                                 0.019377   
fund_return_1year                                   0.048213   
fund_return_3years                                  0.100403   
fund_return_2020                                    0.059120   
fund_return_2019                                    0.053690   
fund_return_2018                                    0.007952   
fund_return_2017                                    0.007643   

                                      Closeness Centrality  
fund_annual_report_net_expense_ratio              0.458730  
fund_price_book_ratio                             0.573413  
fund_price_cashflow_ratio                         0.535185  
fund_price_earning_ratio                          0.535185  
fund_price_sales_ratio                            0.382275  
fund_sharpe_ratio_3years                          0.729798  
fund_treynor_ratio_3years                         0.000000  
fund_beta_3years                                  0.617521  
fund_r_squared_3years                             0.433934  
fund_stdev_3years                                 0.553640  
fund_return_ytd                                   0.594650  
fund_return_1month                                0.535185  
fund_return_3months                               0.594650  
fund_return_1year                                 0.668981  
fund_return_3years                                0.698068  
fund_return_2020                                  0.642222  
fund_return_2019                                  0.642222  
fund_return_2018                                  0.501736  
fund_return_2017                                  0.573413

Going over the centrality results in the network analysis on financial funds and the levels of influence within the network. 

Degree Centrality: Higher degree centrality shows funds are more connected with multiple fund types. This can be influential towards shared movements in returns. fund_return_1year as one example has a high degree of centrality which shows high interconnection and can react similarly to a range of different markets. 

Betweenness Centrality: Funds with high centrality here can control the flow of information and how the spread of influence impacts the networks. It can be critical to know what funds might act as bridges in financial systems during a market crisis. fund_return_1year is an example of this as it has high centrality 

Closeness Centrality: High closeness centrality shows that funds can interact with each other or affect one another in the network. fund_return_1year shows this again as an example, meaning a rapid increase of changes in the market might occur potentially.

Overall for our research topic, funds like fund_return_1year which is a 1-year fund return for ETFs, can play critical roles in the network which can cause systemic risk and market stability. Funds with high betweenness centrality can create predictions on what type of funds would be best used for a crisis that may come up, and keeping an eye on them would be beneficial. 

In [81]:
# To incorporate train and testing model for the centralities, will create a risk level to predict risk level

# Threshold numbers to determine high risk
degree_threshold = 0.5
bet_threshold = 0.1
closeness_threshold = 0.7

# Calculate risk level based on thresholds 
centrality_df['Risk level'] = ((centrality_df['Degree Centrality']> degree_threshold) |
                               (centrality_df['Betweenness Centrality']> bet_threshold) |
                               (centrality_df['Closeness Centrality']> closeness_threshold)).astype(int)
centrality_df

Degree Centrality  \
fund_annual_report_net_expense_ratio           0.111111   
fund_price_book_ratio                          0.388889   
fund_price_cashflow_ratio                      0.333333   
fund_price_earning_ratio                       0.333333   
fund_price_sales_ratio                         0.166667   
fund_sharpe_ratio_3years                       0.777778   
fund_treynor_ratio_3years                      0.111111   
fund_beta_3years                               0.500000   
fund_r_squared_3years                          0.111111   
fund_stdev_3years                              0.333333   
fund_return_ytd                                0.444444   
fund_return_1month                             0.333333   
fund_return_3months                            0.444444   
fund_return_1year                              0.611111   
fund_return_3years                             0.611111   
fund_return_2020                               0.500000   
fund_return_2019                               0.555556   
fund_return_2018                               0.388889   
fund_return_2017                               0.388889   

                                      Betweenness Centrality  \
fund_annual_report_net_expense_ratio                0.000000   
fund_price_book_ratio                               0.058397   
fund_price_cashflow_ratio                           0.027135   
fund_price_earning_ratio                            0.027135   
fund_price_sales_ratio                              0.000000   
fund_sharpe_ratio_3years                            0.179910   
fund_treynor_ratio_3years                           0.000000   
fund_beta_3years                                    0.014070   
fund_r_squared_3years                               0.000000   
fund_stdev_3years                                   0.028994   
fund_return_ytd                                     0.021322   
fund_return_1month                                  0.045985   
fund_return_3months                                 0.019377   
fund_return_1year                                   0.048213   
fund_return_3years                                  0.100403   
fund_return_2020                                    0.059120   
fund_return_2019                                    0.053690   
fund_return_2018                                    0.007952   
fund_return_2017                                    0.007643   

                                      Closeness Centrality  Risk level  
fund_annual_report_net_expense_ratio              0.458730           0  
fund_price_book_ratio                             0.573413           0  
fund_price_cashflow_ratio                         0.535185           0  
fund_price_earning_ratio                          0.535185           0  
fund_price_sales_ratio                            0.382275           0  
fund_sharpe_ratio_3years                          0.729798           1  
fund_treynor_ratio_3years                         0.000000           0  
fund_beta_3years                                  0.617521           0  
fund_r_squared_3years                             0.433934           0  
fund_stdev_3years                                 0.553640           0  
fund_return_ytd                                   0.594650           0  
fund_return_1month                                0.535185           0  
fund_return_3months                               0.594650           0  
fund_return_1year                                 0.668981           1  
fund_return_3years                                0.698068           1  
fund_return_2020                                  0.642222           0  
fund_return_2019                                  0.642222           1  
fund_return_2018                                  0.501736           0  
fund_return_2017                                  0.573413           0

To go over the additional columns and how they influence transaction price, volume, value, loss, and Measure of risk 

Measure of Risk: fund_sharpe_ratio and fund_treynor_ratio are related to risk measurement. Sharpe ratio can help rank and evaluate performance that exceeds the risk-free rate after adjusting the risk. The Treynor ratio is also a risk-adjusted measurement return based on systematic risk. Overall they focus on risk-adjusted returns in funds.

Fund_beta is also a volatility measure that can indicate systematic risk. Fund r squared and Stdev also show a measure of risk. R_squared focuses on fund movements and is explained through movements in a benchmark index, while Stdev focuses on overall volatility to determine risk.

Loss: Hogh centrality levels, specifically betweenness centrality, can determine values that indicate loss areas with a risk level of 1 in our dataset. For example for fund_sharpe_ratio, it has a high betweenness centrality result of 0.17, which may indicate high market risk and can lead to potential loss in the network path. 

Transaction price: Raios such as fund_price_earning_ratio and price_book_ratio are directly involved with transaction price since they focus on specific transaction occurrences to determine profit. Higher centrality specifically in closeness centrality, affects how transactions are valued. 

Transaction Volume: Similar to said above about transaction price, higher centrality in trade volume can show how often it is used in market activity and influence trade behavior in funds. So price_earnings and price_cashflow have high centrality are some values that show these influences. Also, lower values such as price_sales_ratio could lead to a potential increase in buying activity, which could increase volumes.

Measure of risk: fund_sharpe_ratio_3years, fund_treynor_ratio_3years, fund_beta_3years, fund_r_squared_3years, fund_stdev_3years

Loss: Any value with a risk level for 1 

Transaction Price/Volume: fund_price_book_ratio, fund_price_cashflow_ratio, fund_price_earning_ratio, fund_price_sales_ratio

Returns: fund_return_ytd, fund_return_1month, fund_return_3months, fund_return_1year, fund_return_3years, fund_return_2020, fund_return_2019, fund_return_2018, fund_return_2017	